core/dbt/contracts/graph/manifest.py

```diff
@@ -908,6 +957,212 @@ def merge_from_artifact(
+        self.exposures[exposure.unique_id] = exposure
+        self.get_file(source_file).exposures.append(exposure.unique_id)
+
+    def add_disabled_nofile(self, node: CompileResultNode):
+        if node.unique_id in self._disabled:
+            self._disabled[node.unique_id].append(node)
+        else:
+            self._disabled[node.unique_id] = [node]
+
+    def add_disabled(self, source_file: SourceFile, node: CompileResultNode):
+        self.add_disabled_nofile(node)
+        self.get_file(source_file).nodes.append(node.unique_id)
+
+    def add_doc(self, source_file: SourceFile, doc: ParsedDocumentation):
+        _check_duplicates(doc, self.docs)
+        self.docs[doc.unique_id] = doc
+        self.get_file(source_file).docs.append(doc.unique_id)
+
+    def _get_disabled(
+        self,
+        unique_id: str,
+        match_file: SourceFile,
+    ) -> List[CompileResultNode]:
+        if unique_id not in self._disabled:
+            raise InternalException(
+                'called _get_disabled with id={}, but it does not exist'
+                .format(unique_id)
+            )
+        return [
+            n for n in self._disabled[unique_id]
+            if n.original_file_path == match_file.path.original_file_path
+        ]
+
+    # This is only used by 'sanitized_update' which processes "old_manifest"
+    def _process_node(
+        self,
+        node_id: str,
+        source_file: SourceFile,
+        old_file: SourceFile,
+        old_manifest: Any,
+    ) -> None:
+        """Nodes are a special kind of complicated - there can be multiple
+        with the same name, as long as all but one are disabled.
+
+        Only handle nodes where the matching node has the same resource type
+        as the current parser.
+        """
+        source_path = source_file.path.original_file_path
+        found: bool = False
+        if node_id in old_manifest.nodes:
+            old_node = old_manifest.nodes[node_id]
+            if old_node.original_file_path == source_path:
+                self.add_node(source_file, old_node)
+                found = True
+
+        if node_id in old_manifest._disabled:
+            matches = old_manifest._get_disabled(node_id, source_file)
+            for match in matches:
+                self.add_disabled(source_file, match)
+                found = True
+
+        if not found:
+            raise CompilationException(
+                'Expected to find "{}" in cached "manifest.nodes" or '
+                '"manifest.disabled" based on cached file information: {}!'
+                .format(node_id, old_file)
+            )
+
+    # This is called by ManifestLoader._get_cached/parse_with_cache,
+    # which handles updating the ManifestLoader results with information
+    # from the "old_manifest", i.e. the pickle file if the checksums are
+    # the same.
+    def sanitized_update(
+        self,
+        source_file: SourceFile,
+        old_manifest: Any,
```

Comment:
Just a note, you can do something like
```suggestion
        old_manifest: Optional["Manifest"],
```
here to refer to itself as a type.

Reply:
Thanks! I've seen code using the string class name but didn't think of it for here.